src/lib/mina_lib/mina_lib.ml

```diff
@@ -3091,14 +3091,35 @@ module Hardfork_config = struct
       ~next_epoch_ledger_config:(Some genesis_next_epoch_ledger_config)
       ~next_epoch_seed:(Epoch_seed.to_base58_check next_epoch_seed)
 
-  let write_config_file ~filename daemon_config =
-    Async.Writer.save filename
-      ~contents:(Yojson.Safe.to_string (Runtime_config.to_yojson daemon_config))
+  let write_config_file ~staging_path ~final_path daemon_config =
+    let open Deferred.Or_error.Let_syntax in
+    let dir = Filename.dirname final_path in
+    let backup_name = dir ^/ "daemon.berkeley.json" in
+    (* Step 1: Backup existing daemon.json if present *)
+    let%bind () =
+      match%bind.Deferred Sys.file_exists final_path with
+      | `Yes ->
+          let%bind.Deferred contents = Async.Reader.file_contents final_path in
```

Comment:
why is it not just a move?

Also, what happen if the backup file already exist? For example we're running the same commands for 3 times.